src/riak_kv_index_hashtree.erl

```diff
@@ -507,24 +507,29 @@ determine_data_root() ->
 %%      then we immediately use version in capabilities. Otherwise check if capabilities
 %%      has flipped to a version yet and if it has, check if we've already upgraded by
 %%      looking for versioned AAE directory.
--spec determine_version(undefined | list(), non_neg_integer(), list()) -> undefined | non_neg_integer().
+-spec determine_version(undefined | list(), non_neg_integer(), list()) -> legacy | non_neg_integer().
 determine_version(Root, Index, Opts) ->
-    case lists:member(upgrade, Opts) of
-        true ->
-            riak_core_capability:get({riak_kv, object_hash_version}, undefined);
-        false ->
-            case riak_core_capability:get({riak_kv, object_hash_version}, undefined) of
-                V when is_integer(V) ->
-                    %% Check if the v[Version]/Index dir exists
-                    case filelib:is_dir(filename:join(filename:join(Root, "v" ++ integer_to_list(V)),integer_to_list(Index))) of
-                        true ->
-                            V;
-                        false ->
-                            undefined
-                    end;
-                _ ->
-                    undefined
-            end
+    case application:get_env(riak_kv, force_hashtree_upgrade) of
```

Comment:
Please try to break up this 4-level deep case statement into something more manageable.

Take a look at http://www.gar1t.com/blog/solving-embarrassingly-obvious-problems-in-erlang.html for some suggestions as to how you might go about it. I realize there's tons of this in our codebase, just trying to make sure we don't add more of them.